invisible_cities/cities/beersheba.py

```diff
@@ -0,0 +1,330 @@
+    energy_type     : Energy type ('E' or 'Ec', see Esmeralda) used for assignment.
+    deconv_mode     : 'joint' or 'separate', 1 or 2 step deconvolution, see description later.
+    diffusion       : Diffusion coefficients in each dimension for 'separate' mode.
+    n_dim           : Number of dimensions to apply the method (2 or 3).
+    cut_type        : Cut mode to the deconvolution output ('abs' or 'rel') using ecut
+                      'abs': cut on the absolute value of the hits.
+                      'rel': cut on the relative value (to the max) of the hits.
+    interMethod     : Interpolation method.
+
+    Returns
+    ----------
+    apply_deconvolution : Function that takes hits and returns the
+    deconvolved data.
+    """
+
+    psfs          = pd.read_hdf(psf_fname)
+    deconvolution = deconvolve(iterationNumber, iterationThr, sampleWidth[:n_dim], bin_size[:n_dim], interMethod)
+    dimensions    = ['X', 'Y', 'Z'][:n_dim]
+
+    def create_deconvolution_df(hits, deconvE, pos):
+        df = pd.DataFrame(columns=['event', 'npeak', 'X', 'Y', 'Z', 'E'])
+        if   cut_type == 'abs':
+            selDeconv   = deconvE > ecut
+        elif cut_type == 'rel':
+            selDeconv   = deconvE/deconvE.max() > ecut
+        else:
+            raise TypeError(f'Cut type "{cut_type}" unsupported')
+        ene         = deconvE[selDeconv]
+        df['event'] = [hits.event.unique()[0]] * len(ene)
+        df['npeak'] = [hits.npeak.unique()[0]] * len(ene)
+        df['Z']     = [hits.Z    .unique()[0]] * len(ene) if n_dim == 2 else pos[2][selDeconv]
+        df['X']     = pos[0][selDeconv]
+        df['Y']     = pos[1][selDeconv]
+        df['E']     = ene
+
+        return df
+
+    def distribute_energy(df, hdst):
+        df['E'] = df['E'] / df['E'].sum() * hdst[energy_type].sum()
+
+        return df
+
+    def deconvolve_hits(df, z):
+        if   deconv_mode == 'joint':
+            psf = psfs[(psfs.z == find_nearest(psfs.z,                 z)) &
+                       (psfs.x == find_nearest(psfs.x, df.Xpeak.unique())) &
+                       (psfs.y == find_nearest(psfs.y, df.Ypeak.unique())) ]
+            deconvImage, pos = deconvolution(tuple(df[v].values for v in dimensions), df.Q.values, psf)
+        elif deconv_mode == 'separate':
+            psf_z0 = psfs[(psfs.z == find_nearest(psfs.z,                 0)) &
+                          (psfs.x == find_nearest(psfs.x, df.Xpeak.unique())) &
+                          (psfs.y == find_nearest(psfs.y, df.Ypeak.unique())) ]
+            deconvImage, pos = deconvolution(tuple(df[v].values for v in dimensions), df.Q.values, psf_z0)
+            g   = multivariate_normal(tuple(0. for _ in range(n_dim)), tuple((diff * sqrt(z/10))**2 for diff in diffusion[:n_dim])).pdf(list(zip(*tuple(psf_z0[f'{v.lower()}r'].values for v in dimensions))))
+            psf = g.reshape(tuple(psf_z0[f'{v.lower()}r'].nunique() for v in dimensions))
```

Comment:
following previous comment...
```suggestion
            psf = gaus.reshape(psf_cols.nunique())
```